GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/IncludedFolderTests.cs

```diff
@@ -0,0 +1,262 @@
+        [TestCase, Order(6)]
+        public void ReadFileThenChangingIncludeFoldersShouldRemoveFileAndFolder()
+        {
+            string fileToRead = Path.Combine(this.Enlistment.RepoRoot, "Scripts", "RunFunctionalTests.bat");
+            this.fileSystem.ReadAllText(fileToRead);
+
+            this.gvfsProcess.AddIncludedFolders(this.mainIncludedFolder);
+            this.ValidateIncludedFolders(this.mainIncludedFolder);
+
+            string folderPath = Path.Combine(this.Enlistment.RepoRoot, "Scripts");
+            folderPath.ShouldNotExistOnDisk(this.fileSystem);
+            fileToRead.ShouldNotExistOnDisk(this.fileSystem);
+        }
+
+        [TestCase, Order(7)]
+        public void CreateNewFileWillPreventRemoveIncludedFolder()
+        {
+            this.gvfsProcess.AddIncludedFolders(this.mainIncludedFolder, "Scripts");
+            this.ValidateIncludedFolders(this.mainIncludedFolder, "Scripts");
+
+            string fileToCreate = Path.Combine(this.Enlistment.RepoRoot, "Scripts", "newfile.txt");
+            this.fileSystem.WriteAllText(fileToCreate, "New Contents");
+
+            this.gvfsProcess.RemoveIncludedFolders("Scripts");
+            this.ValidateIncludedFolders(this.mainIncludedFolder, "Scripts");
+
+            string folderPath = Path.Combine(this.Enlistment.RepoRoot, "Scripts");
+            folderPath.ShouldBeADirectory(this.fileSystem);
+            string[] fileSystemEntries = Directory.GetFileSystemEntries(folderPath);
+            fileSystemEntries.Length.ShouldEqual(6);
+            fileToCreate.ShouldBeAFile(this.fileSystem);
+
+            this.fileSystem.DeleteFile(fileToCreate);
+        }
+
+        [TestCase, Order(8)]
+        public void ModifiedFileShouldNotAllowIncludedFolderChange()
+        {
+            string modifiedPath = Path.Combine(this.Enlistment.RepoRoot, "Scripts", "RunFunctionalTests.bat");
+            this.fileSystem.WriteAllText(modifiedPath, "New Contents");
+
+            string output = this.gvfsProcess.AddIncludedFolders(this.mainIncludedFolder);
+            output.ShouldContain("Include was aborted");
+            this.ValidateIncludedFolders(new string[0]);
+        }
+
+        [TestCase, Order(9)]
+        public void ModifiedFileAndCommitThenChangingIncludeFoldersShouldKeepFileAndFolder()
+        {
+            string modifiedPath = Path.Combine(this.Enlistment.RepoRoot, "Scripts", "RunFunctionalTests.bat");
+            this.fileSystem.WriteAllText(modifiedPath, "New Contents");
+            GitProcess.Invoke(this.Enlistment.RepoRoot, "add .");
+            GitProcess.Invoke(this.Enlistment.RepoRoot, "commit -m Test");
+
+            this.gvfsProcess.AddIncludedFolders(this.mainIncludedFolder);
+            this.ValidateIncludedFolders(this.mainIncludedFolder);
+
+            string folderPath = Path.Combine(this.Enlistment.RepoRoot, "Scripts");
+            folderPath.ShouldBeADirectory(this.fileSystem);
+            modifiedPath.ShouldBeAFile(this.fileSystem);

```

Comment:
Why were these kept? Is it because the file is now a full file?

What would happen if the file were deleted instead of modified? (it might be useful to add a test for that as well)

Reply:
Yes it is because files that get added to ModifiedPaths.dat will stay there and git is in charge of them even if they are outside the sparse scope.  Added a delete test as well in which case the file and folder are gone but if you checkout a commit with the file then it will appear when outside the scope.